MAINT: Clean and update documentation

HaoZeke opened the pull request:
Closes #1427 and re-hauls the documentation for the new release. For the new release.

mattip commented:
This is a very large PR, I will find it hard to review. Is there something in particular you would like me to look at?

HaoZeke commented:
> This is a very large PR, I will find it hard to review. Is there something in particular you would like me to look at?

Ah sorry, I'll highlight a few places, in general there aren't too many differences, mostly cosmetic / documentation changes, that is:
- Uses `bibtex` for references, adds a logo (light/dark) and switched to `furo`
- Removed the spurious makefile stuff (which accounts for most of the deletions)
- Moved long module documentation (step detection) into its own page
- Uses `AutoAPI` and `viewcode` for more through documentation
  - This also means the `sphinx` build doesn't import `asv` anymore 
  - Uses `importlib_metadata` to consistently get the version as suggested by `setuptools_scm`

Here is what the documentation looks like at this commit:
https://tmpasv.surge.sh/

mattip commented:
It all seems quite straightforward, go ahead an merge if you are satisfied.
